core/src/main/java/org/elasticsearch/index/reindex/Slices.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Represents the setting for the number of slices used by a BulkByScrollRequest. Valid values are positive integers and "auto". The
+ * "auto" setting defers choosing the number of slices to the request handler.
+ */
+public final class Slices implements ToXContent, Writeable {
+
+    private static final int AUTO_COUNT = -1;
+
+    public static final String FIELD_NAME = "slices";
+    public static final String AUTO_VALUE = "auto";
+
+    public static final Slices AUTO = new Slices(AUTO_COUNT);
+    public static final Slices ONE = of(1);
+    public static final Slices DEFAULT = ONE;
+
+    private final int count;
+
+    private Slices(int count) {
+        this.count = count;
+    }
+
+    public Slices(StreamInput stream) throws IOException {
```

Comment:
I like to put the reading code right under the writing code so they are easy to eyeball together.